File: app/src/main/java/com/absinthe/anywhere_/viewmodel/AnywhereViewModel.java
package com.absinthe.anywhere_.viewmodel;

import android.app.Activity;
import android.app.Application;
import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.Build;

import androidx.fragment.app.Fragment;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import com.absinthe.anywhere_.AnywhereApplication;
import com.absinthe.anywhere_.R;
import com.absinthe.anywhere_.adapter.page.PageNode;
import com.absinthe.anywhere_.adapter.page.PageTitleNode;
import com.absinthe.anywhere_.database.AnywhereRepository;
import com.absinthe.anywhere_.model.AnywhereEntity;
import com.absinthe.anywhere_.model.AnywhereType;
import com.absinthe.anywhere_.model.Const;
import com.absinthe.anywhere_.model.GlobalValues;
import com.absinthe.anywhere_.model.PageEntity;
import com.absinthe.anywhere_.services.CollectorService;
import com.absinthe.anywhere_.utils.AppUtils;
import com.absinthe.anywhere_.utils.ToastUtil;
import com.absinthe.anywhere_.utils.manager.Logger;
import com.absinthe.anywhere_.utils.manager.ShizukuHelper;
import com.absinthe.anywhere_.view.editor.Editor;
import com.absinthe.anywhere_.view.editor.ImageEditor;
import com.absinthe.anywhere_.view.editor.SchemeEditor;
import com.absinthe.anywhere_.view.editor.ShellEditor;
import com.blankj.utilcode.util.DeviceUtils;
import com.blankj.utilcode.util.PermissionUtils;
import com.chad.library.adapter.base.entity.node.BaseNode;

import java.util.ArrayList;
import java.util.List;

public class AnywhereViewModel extends AndroidViewModel {

    private AnywhereRepository mRepository;
    private LiveData<List<AnywhereEntity>> mAllAnywhereEntities;

    private MutableLiveData<String> mBackground = new MutableLiveData<>();
    private MutableLiveData<Fragment> mFragment = new MutableLiveData<>();

    public AnywhereViewModel(Application application) {
        super(application);
        mRepository = AnywhereApplication.sRepository;
        mAllAnywhereEntities = mRepository.getAllAnywhereEntities();
    }

    public LiveData<List<AnywhereEntity>> getAllAnywhereEntities() {
        return mAllAnywhereEntities;
    }

    public void insert(AnywhereEntity ae) {
        mRepository.insert(ae);
    }

    public void update(AnywhereEntity ae) {
        mRepository.update(ae);
    }

    public void delete(AnywhereEntity ae) {
        mRepository.delete(ae);
    }

    public MutableLiveData<String> getBackground() {
        if (mBackground == null) {
            mBackground = new MutableLiveData<>();
        }
        return mBackground;
    }

    public MutableLiveData<Fragment> getFragment() {
        if (mFragment == null) {
            mFragment = new MutableLiveData<>();
        }
        return mFragment;
    }

    public PageTitleNode getEntity(String title) {
        List<BaseNode> pageNodeList = new ArrayList<>();
        PageNode pageNode = new PageNode();
        pageNode.setTitle(title);
        pageNodeList.add(pageNode);
        PageTitleNode pageTitle = new PageTitleNode(pageNodeList, title);
        pageTitle.setExpanded(title.equals(GlobalValues.sCategory));
        return pageTitle;
    }

    public void setUpUrlScheme(Context context, String url) {
        AnywhereEntity ae = AnywhereEntity.Builder();
        ae.setAppName(getApplication().getString(R.string.bsd_new_url_scheme_name));
        ae.setParam1(url);
        ae.setType(AnywhereType.URL_SCHEME);

        Editor editor = new SchemeEditor(context)
                .item(ae)
                .isEditorMode(false)
                .isShortcut(false)
                .build();
        editor.show();
    }

    public void setUpUrlScheme(Context context) {
        setUpUrlScheme(context, "");
    }

    public void openImageEditor(Context context, boolean isDismissParent) {
        AnywhereEntity ae = AnywhereEntity.Builder();
        ae.setAppName("New Image");
        ae.setType(AnywhereType.IMAGE);

        Editor editor = new ImageEditor(context)
                .item(ae)
                .isEditorMode(false)
                .isShortcut(false)
                .setDismissParent(isDismissParent)
                .build();
        editor.show();
    }

    public void openShellEditor(Context context, boolean isDismissParent) {
        AnywhereEntity ae = AnywhereEntity.Builder();
        ae.setAppName("New Shell");
        ae.setType(AnywhereType.SHELL);

        Editor editor = new ShellEditor(context)
                .item(ae)
                .isEditorMode(false)
                .isShortcut(false)
                .setDismissParent(isDismissParent)
                .build();
        editor.show();
    }

    public void startCollector(Activity activity) {
        switch (GlobalValues.getWorkingMode()) {
            case Const.WORKING_MODE_URL_SCHEME:
                setUpUrlScheme(activity);
                break;
            case Const.WORKING_MODE_SHIZUKU:
                if (Build.VERSION.SDK_INT < Build.VERSION_CODES.M) {
                    if (ShizukuHelper.checkShizukuOnWorking(activity) && ShizukuHelper.isGrantShizukuPermission()) {
                        CollectorService.startCollector(activity);
                    } else {
                        ShizukuHelper.requestShizukuPermission();
                    }
                } else {
                    PermissionUtils.requestDrawOverlays(new PermissionUtils.SimpleCallback() {
                        @Override
                        public void onGranted() {
                            if (ShizukuHelper.checkShizukuOnWorking(activity) && ShizukuHelper.isGrantShizukuPermission()) {
                                CollectorService.startCollector(activity);
                            } else {
                                ShizukuHelper.requestShizukuPermission();
                            }
                        }

                        @Override
                        public void onDenied() {

                        }
                    });
                }
                break;
            case Const.WORKING_MODE_ROOT:
                if (Build.VERSION.SDK_INT < Build.VERSION_CODES.M) {
                    if (DeviceUtils.isDeviceRooted()) {
                        CollectorService.startCollector(activity);
                    } else {
                        Logger.d("ROOT permission denied.");
                        ToastUtil.makeText(R.string.toast_root_permission_denied);
                        com.absinthe.anywhere_.utils.PermissionUtils.upgradeRootPermission(activity.getPackageCodePath());
                    }
                } else {
                    PermissionUtils.requestDrawOverlays(new PermissionUtils.SimpleCallback() {
                        @Override
                        public void onGranted() {
                            if (DeviceUtils.isDeviceRooted()) {
                                CollectorService.startCollector(activity);
                            } else {
                                Logger.d("ROOT permission denied.");
                                ToastUtil.makeText(R.string.toast_root_permission_denied);
                                com.absinthe.anywhere_.utils.PermissionUtils.upgradeRootPermission(activity.getPackageCodePath());
                            }
                        }

                        @Override
                        public void onDenied() {

                        }
                    });
                }
                break;
            default:
        }
    }

    public void addPage() {
        List<PageEntity> list = mRepository.getAllPageEntities().getValue();
        if (list != null) {
            if (list.size() != 0) {
                int size = list.size();
                PageEntity pe = PageEntity.Builder();
                pe.setTitle("Page " + (size + 1));
                pe.setPriority(size + 1);
                pe.setType(AnywhereType.CARD_PAGE);
                mRepository.insertPage(pe);
            } else {
                PageEntity pe = PageEntity.Builder();
                pe.setTitle(AnywhereType.DEFAULT_CATEGORY);
                pe.setPriority(1);
                pe.setType(AnywhereType.CARD_PAGE);
                mRepository.insertPage(pe);
            }
        }
    }

    public void addWebPage(Uri uri, Intent intent) {
        List<PageEntity> list = mRepository.getAllPageEntities().getValue();
        if (list != null) {
            int size = list.size();
            PageEntity pe = PageEntity.Builder();
            pe.setTitle("Web Page " + (size + 1));
            pe.setPriority(size + 1);
            pe.setType(AnywhereType.WEB_PAGE);
            pe.setExtra(uri.toString());
            mRepository.insertPage(pe);
            AppUtils.takePersistableUriPermission(getApplication(), uri, intent);
        }
    }
}

File: app/src/main/java/com/absinthe/anywhere_/viewmodel/AnywhereViewModel.kt
package com.absinthe.anywhere_.viewmodel

import android.app.Activity
import android.app.Application
import android.content.Context
import android.content.Intent
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import com.absinthe.anywhere_.AnywhereApplication
import com.absinthe.anywhere_.R
import com.absinthe.anywhere_.adapter.page.PageNode
import com.absinthe.anywhere_.adapter.page.PageTitleNode
import com.absinthe.anywhere_.constants.AnywhereType
import com.absinthe.anywhere_.constants.Const
import com.absinthe.anywhere_.constants.GlobalValues
import com.absinthe.anywhere_.database.AnywhereRepository
import com.absinthe.anywhere_.model.database.AnywhereEntity
import com.absinthe.anywhere_.model.database.PageEntity
import com.absinthe.anywhere_.ui.editor.EXTRA_EDIT_MODE
import com.absinthe.anywhere_.ui.editor.EXTRA_ENTITY
import com.absinthe.anywhere_.ui.editor.EditorActivity
import com.absinthe.anywhere_.utils.AppUtils
import com.absinthe.anywhere_.utils.ToastUtil
import com.absinthe.anywhere_.utils.manager.ShellManager
import com.absinthe.anywhere_.utils.manager.ShizukuHelper
import com.blankj.utilcode.util.DeviceUtils
import com.blankj.utilcode.util.PermissionUtils
import com.chad.library.adapter.base.entity.node.BaseNode
import com.topjohnwu.superuser.Shell
import timber.log.Timber

class AnywhereViewModel(application: Application) : AndroidViewModel(application) {

  private val mRepository: AnywhereRepository = AnywhereApplication.sRepository
  val allAnywhereEntities: LiveData<List<AnywhereEntity>> = mRepository.allAnywhereEntities
  var shouldShowFab: MutableLiveData<Boolean> = MutableLiveData()
  var background: MutableLiveData<String> = MutableLiveData()
    private set

  fun insert(ae: AnywhereEntity) {
    mRepository.insert(ae)
  }

  fun update(ae: AnywhereEntity) {
    mRepository.update(ae)
  }

  fun delete(ae: AnywhereEntity) {
    mRepository.delete(ae)
  }

  fun getEntity(title: String): PageTitleNode {
    val pageNodeList: MutableList<BaseNode> = ArrayList()

    pageNodeList.add(PageNode().apply {
      this.title = title
    })

    return PageTitleNode(pageNodeList, title).apply {
      isExpanded = (title == GlobalValues.category)
    }
  }

  fun setUpUrlScheme(context: Context, url: String = "") {
    val ae = AnywhereEntity().apply {
      appName = AnywhereType.Card.NEW_TITLE_MAP[AnywhereType.Card.URL_SCHEME]!!
      param1 = url
      type = AnywhereType.Card.URL_SCHEME
    }
    context.startActivity(Intent(context, EditorActivity::class.java).apply {
      putExtra(EXTRA_ENTITY, ae)
      putExtra(EXTRA_EDIT_MODE, false)
    })
  }

  fun startCollector(activity: Activity, listener: OnStartCollectorListener) {
    when (GlobalValues.workingMode) {
      Const.WORKING_MODE_URL_SCHEME -> {
        ToastUtil.makeText(R.string.toast_works_on_root_or_shizuku)
      }
      Const.WORKING_MODE_SHIZUKU -> {
        if (PermissionUtils.isGrantedDrawOverlays()) {
          if (ShizukuHelper.checkPermission(activity)) {
            listener.onStart()
          }
        } else {
          if (AppUtils.atLeastR()) {
            ToastUtil.makeText(R.string.toast_overlay_choose_anywhere)
          }
          PermissionUtils.requestDrawOverlays(object : PermissionUtils.SimpleCallback {
            override fun onGranted() {
              if (ShizukuHelper.checkPermission(activity)) {
                listener.onStart()
              }
            }

            override fun onDenied() {}
          })
        }
      }
      Const.WORKING_MODE_ROOT -> {
        if (PermissionUtils.isGrantedDrawOverlays()) {
          if (Shell.isAppGrantedRoot() == true) {
            listener.onStart()
          } else {
            Timber.d("ROOT permission denied.")
            ToastUtil.makeText(R.string.toast_root_permission_denied)
            ShellManager.acquireRoot()
          }
        } else {
          if (AppUtils.atLeastR()) {
            ToastUtil.makeText(R.string.toast_overlay_choose_anywhere)
          }
          PermissionUtils.requestDrawOverlays(object : PermissionUtils.SimpleCallback {
            override fun onGranted() {
              if (DeviceUtils.isDeviceRooted()) {
                listener.onStart()
              } else {
                Timber.d("ROOT permission denied.")
                ToastUtil.makeText(R.string.toast_root_permission_denied)
                ShellManager.acquireRoot()
              }
            }

            override fun onDenied() {}
          })
        }
      }
    }
  }

  fun addPage() {
    mRepository.allPageEntities.value?.let { pages ->
      val pe = PageEntity().apply {
        if (pages.isNotEmpty()) {
          var count = 1
          var t = "Page " + (pages.size + count++)
          while (pages.any { it.title == t }) {
            t = "Page " + (pages.size + count++)
          }
          title = t
          priority = pages.size + 1
        } else {
          title = AnywhereType.Category.DEFAULT_CATEGORY
          priority = 1
        }
        type = AnywhereType.Page.CARD_PAGE
      }
      mRepository.insertPage(pe)
    }
  }

  interface OnStartCollectorListener {
    fun onStart()
  }
}
